# illegal-break-continue (QAS009)

**Severity:** error  
**Category:** qasm3, control-flow, syntax  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/classical.html#breaking-and-continuing-loops)  

## Description

break and continue can only be used inside loops.

## Rule Details

This rule checks for illegal break continue violations according to OpenQASM 3.0 specifications.

## Message Format

```
'{{ keyword }}' cannot be used outside of a loop.
```

## Examples

### ❌ Incorrect

```qasm
break;  // outside of loop, error
```

### ✅ Correct

```qasm
for i in [0:1:10] {
  if (i == 5) break;  // inside loop, valid
}
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** break_or_continue

## Related Rules

- [QAS010](QAS010.md) (invalid-instruction-in-gate): Both relate to syntax constraints
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/classical.html#breaking-and-continuing-loops)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS009.md)
