# unused-qubit (QAS001)

**Severity:** warning  
**Category:** qasm3, readability, unused-variables  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/types.html#qubits)  

## Description

Detects qubits that are declared but never used in gates or measurements.

## Rule Details

This rule checks for unused qubit violations according to OpenQASM 3.0 specifications.

## Message Format

```
Qubit '{{ name }}' is declared but never used.
```

## Examples

### ❌ Incorrect

```qasm
qubit q;
qubit unused_qubit;  // Never used
h q;
```

### ✅ Correct

```qasm
qubit q;
h q;  // Qubit is used
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** qubit

## Related Rules

- [QAS005](QAS005.md) (naming-convention-violation): Both relate to declaration best practices
- [QAS008](QAS008.md) (qubit-declared-in-local-scope): Both relate to qubit declaration
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/types.html#qubits)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS001.md)
