# gate-parameter-indexing (QAS007)

**Severity:** error  
**Category:** qasm3, gate, syntax  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/gates.html#hierarchical-gates-definitions)  

## Description

Index access on registers passed as gate arguments is not allowed in gate body.

## Rule Details

This rule checks for gate parameter indexing violations according to OpenQASM 3.0 specifications.

## Message Format

```
Cannot perform index access on gate argument '{{ name }}'.
```

## Examples

### ❌ Incorrect

```qasm
gate mygate(a) {
  h a[0];  // prohibited index access
}
```

### ✅ Correct

```qasm
gate mygate(a) {
  h a;  // use parameter directly
}
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** gate_body

## Related Rules

- [QAS006](QAS006.md) (gate-register-size-mismatch): Both relate to gate parameter handling
- [QAS010](QAS010.md) (invalid-instruction-in-gate): Both relate to gate definition constraints
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/gates.html#hierarchical-gates-definitions)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS007.md)
