# naming-convention-violation (QAS005)

**Severity:** warning  
**Category:** qasm3, style, naming, readability  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)  

## Description

Identifier name violates OpenQASM naming conventions.

## Rule Details

This rule checks for naming convention violation violations according to OpenQASM 3.0 specifications.

## Message Format

```
Identifier '{{ name }}' violates naming conventions. Follow pattern: ^[a-z][a-zA-Z0-9_]*$.
```

## Examples

### ❌ Incorrect

```qasm
qubit[2] MyQubit;  // starts with uppercase
```

### ✅ Correct

```qasm
qubit[2] myQubit;  // starts with lowercase
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** any

## Related Rules

- [QAS001](QAS001.md) (unused-qubit): Both relate to declaration best practices
- [QAS012](QAS012.md) (snake-case-required): Both relate to naming conventions
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS005.md)
