# gate-register-size-mismatch (QAS006)

**Severity:** error  
**Category:** qasm3, semantic, gate  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/gates.html#broadcasting)  

## Description

Quantum register sizes do not match during gate call, preventing broadcasting.

## Rule Details

This rule checks for gate register size mismatch violations according to OpenQASM 3.0 specifications.

## Message Format

```
Register lengths passed to gate '{{ gate }}' do not match.
```

## Examples

### ❌ Incorrect

```qasm
gate mygate(a, b) { ... }
qubit[2] q1;
qubit[3] q2;
mygate q1, q2;  // size mismatch: 2 vs 3
```

### ✅ Correct

```qasm
gate mygate(a, b) { ... }
qubit[2] q1;
qubit[2] q2;
mygate q1, q2;  // both size 2
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** gate_call

## Related Rules

- [QAS004](QAS004.md) (out-of-bounds-index): Both relate to size validation
- [QAS007](QAS007.md) (gate-parameter-indexing): Both relate to gate parameter handling
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/gates.html#broadcasting)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS006.md)
