# constant-measured-bit (QAS003)

**Severity:** warning  
**Category:** qasm3, logic, measurement, constant-results  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/quantum.html#measurement)  

## Description

Measuring a qubit that has no gates applied. The result will always be |0⟩.

## Rule Details

This rule checks for constant measured bit violations according to OpenQASM 3.0 specifications.

## Message Format

```
Measuring qubit '{{ name }}' that has no gates applied. The result will always be |0⟩.
```

## Examples

### ❌ Incorrect

```qasm
qubit q;
bit c;
measure q -> c;  // q has no gates applied
```

### ✅ Correct

```qasm
qubit q;
bit c;
h q;  // apply gate first
measure q -> c;
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** measure

## Related Rules

- [QAS010](QAS010.md) (invalid-instruction-in-gate): Both relate to measurement operations
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/quantum.html#measurement)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS003.md)
