# naming-convention-violation (QAS005)

**Severity:** warning  
**Category:** qasm3, style, naming, readability  
**Fixable:** false  

## Description

Detects violations of OpenQASM naming conventions for variables and circuits.

## Rule Details

This rule checks for naming-convention-violation violations.

## Message Format

```
Identifier '{{ name }}' violates naming convention. Should match pattern: ^[a-z][a-zA-Z0-9_]*$
```

## Examples

### ❌ Incorrect

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit MyQubit;  // ❌ Should start with lowercase
qubit q_2;      // ❌ Should not start with uppercase
```
### ✅ Correct

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit myQubit;  // ✅ Starts with lowercase
qubit q2;       // ✅ Valid naming convention
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** any

## Related Rules

- QAS001 (unused-qubit): Both relate to variable declaration best practices
