# out-of-bounds-index (QAS004)

**Severity:** error  
**Category:** qasm3, bounds-checking, array-access  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/types.html#index-sets-and-slicing)  

## Description

Index of qubit, bit, or array exceeds the defined size.

## Rule Details

This rule checks for out of bounds index violations according to OpenQASM 3.0 specifications.

## Message Format

```
Index out of bounds: accessing '{{ index }}' on '{{ name }}' of length {{ length }}.
```

## Examples

### ❌ Incorrect

```qasm
qubit[2] q;
h q[2];  // q[2] does not exist
```

### ✅ Correct

```qasm
qubit[2] q;
h q[1];  // valid index
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** any

## Related Rules

- [QAS002](QAS002.md) (undefined-identifier): Both relate to identifier validation
- [QAS006](QAS006.md) (gate-register-size-mismatch): Both relate to array bounds checking
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/types.html#index-sets-and-slicing)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS004.md)
