# constant-measured-bit (QAS003)

**Severity:** warning  
**Category:** qasm3, logic, measurement, constant-results  
**Fixable:** false  

## Description

Detects measurements of qubits that have not been affected by any gates, resulting in constant measurement outcomes.

## Rule Details

This rule checks for constant-measured-bit violations.

## Message Format

```
Measuring qubit '{{ name }}' that has not been affected by any gates will always yield |0⟩.
```

## Examples

### ❌ Incorrect

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit q;
bit c;

measure q -> c;  // ❌ Measuring unaffected qubit
```
### ✅ Correct

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit q;
bit c;

h q;          // ✅ Affect qubit before measurement
measure q -> c;
```## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** measure

## Related Rules

- QAS002 (insufficient-classical-bits): Both relate to measurement operations
