# invalid-instruction-in-gate (QAS010)

**Severity:** error  
**Category:** qasm3, gate, syntax  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/gates.html#hierarchical-gates-definitions)  

## Description

Using non-unitary instructions (measurement, reset, etc.) within gate definition.

## Rule Details

This rule checks for invalid instruction in gate violations according to OpenQASM 3.0 specifications.

## Message Format

```
Invalid instruction '{{ stmt }}' used within gate definition.
```

## Examples

### ❌ Incorrect

```qasm
gate g(a) {
  measure a -> c;  // measurement is non-unitary, not allowed
}
```

### ✅ Correct

```qasm
gate g(a) {
  h a;  // unitary operations only
}
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** gate_body

## Related Rules

- [QAS003](QAS003.md) (constant-measured-bit): Both relate to measurement operations
- [QAS007](QAS007.md) (gate-parameter-indexing): Both relate to gate definition constraints
- [QAS009](QAS009.md) (illegal-break-continue): Both relate to syntax constraints
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/gates.html#hierarchical-gates-definitions)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS010.md)
