# qubit-declared-in-local-scope (QAS008)

**Severity:** error  
**Category:** qasm3, scope, qubit  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/types.html#qubits)  

## Description

Qubits can only be declared in global scope.

## Rule Details

This rule checks for qubit declared in local scope violations according to OpenQASM 3.0 specifications.

## Message Format

```
Qubit '{{ name }}' can only be declared in global scope.
```

## Examples

### ❌ Incorrect

```qasm
def foo() {
  qubit q;  // cannot declare in local scope
}
```

### ✅ Correct

```qasm
qubit q;  // global scope declaration
def foo() {
  // use q here
}
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** qubit

## Related Rules

- [QAS001](QAS001.md) (unused-qubit): Both relate to qubit management
- [QAS002](QAS002.md) (undefined-identifier): Both relate to scope validation
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/types.html#qubits)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS008.md)
