# snake-case-required (QAS012)

**Severity:** warning  
**Category:** qasm3, style, naming  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)  

## Description

Identifiers should be named in snake_case (lower_snake_case).

## Rule Details

This rule checks for snake case required violations according to OpenQASM 3.0 specifications.

## Message Format

```
Identifier '{{ name }}' should be written in snake_case.
```

## Examples

### ❌ Incorrect

```qasm
qubit myQubit;  // camelCase
gate MyGate(q) { h q; }  // PascalCase
```

### ✅ Correct

```qasm
qubit my_qubit;  // snake_case
gate my_gate(q) { h q; }  // snake_case
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** any

## Related Rules

- [QAS005](QAS005.md) (naming-convention-violation): Both relate to naming conventions
- [QAS011](QAS011.md) (reserved-prefix-usage): Both relate to naming standards
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS012.md)
