# exceeding-qubit-limits (QAS004)

**Severity:** error  
**Category:** qasm3, bounds-checking, array-access  
**Fixable:** false  

## Description

Detects when qubit or classical bit array indices exceed declared bounds.

## Rule Details

This rule checks for exceeding-qubit-limits violations.

## Message Format

```
Array index out of bounds: {{ issue }}
```

## Examples

### ❌ Incorrect

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit[2] q;

h q[3];  // ❌ Index 3 exceeds array bounds [0,1]
```
### ✅ Correct

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit[4] q;

h q[3];  // ✅ Index 3 is within bounds [0,3]
```

## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** any

## Related Rules

- QAS002 (insufficient-classical-bits): Both relate to array bounds checking
