# undefined-identifier (QAS002)

**Severity:** error  
**Category:** qasm3, scope, semantic  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/scope.html)  

## Description

Identifier is used without being declared.

## Rule Details

This rule checks for undefined identifier violations according to OpenQASM 3.0 specifications.

## Message Format

```
Identifier '{{ name }}' is not declared.
```

## Examples

### ❌ Incorrect

```qasm
h q[0];  // q is undefined
```

### ✅ Correct

```qasm
qubit q;
h q[0];
```

## Configuration

- **Enabled by default:** true
- **Match type:** usage
- **Match kind:** identifier

## Related Rules

- [QAS004](QAS004.md) (out-of-bounds-index): Both relate to identifier usage validation
- [QAS008](QAS008.md) (qubit-declared-in-local-scope): Both relate to scope management
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/scope.html)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS002.md)
