# reserved-prefix-usage (QAS011)

**Severity:** error  
**Category:** qasm3, naming, style  
**Fixable:** false  
**OpenQASM Specification:** [View Details](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)  

## Description

Using reserved prefix (__) in identifier.

## Rule Details

This rule checks for reserved prefix usage violations according to OpenQASM 3.0 specifications.

## Message Format

```
Identifier '{{ name }}' uses reserved prefix '__'.
```

## Examples

### ❌ Incorrect

```qasm
qubit __reserved_name;  // uses reserved prefix
```

### ✅ Correct

```qasm
qubit my_qubit;  // normal identifier
```

## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** any

## Related Rules

- [QAS005](QAS005.md) (naming-convention-violation): Both relate to identifier naming
- [QAS012](QAS012.md) (snake-case-required): Both relate to naming standards
## References

- [OpenQASM 3.0 Specification](https://openqasm.com/versions/3.0/language/lexical.html#identifiers)
- [Rule Documentation](https://github.com/orangekame3/qasmtools/blob/main/docs/rules/QAS011.md)
