# unused-qubit (QAS001)

**Severity:** warning  
**Category:** qasm3, readability, unused-variables  
**Fixable:** false  

## Description

Detects qubits that are declared but never used in gates or measurements.

## Rule Details

This rule checks for unused-qubit violations.

## Message Format

```
Qubit '{{ name }}' is declared but never used.
```

## Examples

### ❌ Incorrect

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit q;
qubit unused_qubit;  // ❌ Never used

h q;
```
### ✅ Correct

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit q;

h q;  // ✅ Qubit is used
```## Configuration

- **Enabled by default:** true
- **Match type:** declaration
- **Match kind:** qubit

## Related Rules

- QAS005 (naming-convention-violation): Both relate to variable declaration best practices
