# insufficient-classical-bits (QAS002)

**Severity:** error  
**Category:** qasm3, correctness, measurement  
**Fixable:** false  

## Description

Detects when the number of classical bits is insufficient for measurements.

## Rule Details

This rule checks for insufficient-classical-bits violations.

## Message Format

```
Insufficient classical bits for measurements. Need {{ required }} but only {{ available }} declared.
```

## Examples

### ❌ Incorrect

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit[2] q;
bit c;  // ❌ Only 1 bit for 2 measurements

h q;
measure q -> c;
```
### ✅ Correct

```qasm
OPENQASM 3.0;
include "stdgates.qasm";

qubit[2] q;
bit[2] c;  // ✅ Sufficient bits for measurements

h q;
measure q -> c;
```## Configuration

- **Enabled by default:** true
- **Match type:** statement
- **Match kind:** measure

## Related Rules

- QAS003 (constant-measured-bit): Both relate to measurement operations
- QAS004 (exceeding-qubit-limits): Both relate to array bounds checking
